### PR TITLE
 Oracle implementation for the CoinMarketCap source

### LINKED ISSUE #4: Add coinmarketcap.com as source for teeracle

### PR DESCRIPTION
**Pull request by murerfel**
As described in #4, implement CoinMarketCap as source. This branch/PR is based on @echevrier 's PR #13, but merged with the new refactoring and abstraction of the oracle logic

- [x] Merge PR #23 first

Closes #4

### CODE REVIEW

File: app-libs/exchange-oracle/src/coin_market_cap.rs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+const FIAT_CURRENCY_PARAM: &str = "convert_id";
+const CRYPTO_CURRENCY_PARAM: &str = "id";
+const COINMARKETCAP_PATH: &str = "v2/cryptocurrency/quotes/latest"; //For free plan
+const COINMARKETCAP_TIMEOUT: Duration = Duration::from_secs(3u64);
+
+lazy_static! {
+	static ref CRYPTO_SYMBOL_ID_MAP: HashMap<&'static str, &'static str> =
+		HashMap::from([("DOT", "6636"), ("TEER", "13323"), ("KSM", "5034"), ("BTC", "1"),]);
+	static ref COINMARKETCAP_KEY: String =
+		env::var("COINMARKETCAP_KEY").unwrap_or_else(|_| "".to_string());
```
Comment (murerfel) [inline app-libs/exchange-oracle/src/coin_market_cap.rs]: Is there not a `unwrap_or_default()` method for this?

File: app-libs/exchange-oracle/src/coin_market_cap.rs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+			)
+			.map_err(Error::RestClient)?;
+
+		let data_struct = response.0;
+
+		let data = match data_struct.data.get(&crypto_id) {
+			Some(d) => d,
+			None => {
+				error!("Got no market data from CoinMarketCap for {} ", crypto_id);
+				return Err(Error::NoValidData)
```
Comment (murerfel) [inline app-libs/exchange-oracle/src/coin_market_cap.rs]: I just noticed this now - I think this is not the standard pattern to handle errors. We should use error types that contain all the relevant information (`NoValidData` should take a String, or whatever type `crypto_id` is and have a corresponding description with `#[error(..)]`). The caller is then responsible for handling the error and deciding if the error is just forwarded, or handled (handling can include writing it to log)

File: enclave-runtime/src/lib.rs [hunk 3]
```diff
@@ -687,18 +691,45 @@ fn update_market_data_internal(
...
 	// Get the exchange rate
 	let trading_pair = TradingPair { crypto_currency, fiat_currency };
-	let coingecko_client = create_coingecko_oracle(Arc::new(OcallApi));
-	let (rate, base_url) = match coingecko_client.get_exchange_rate(trading_pair.clone()) {
-		Ok(r) => r,
+
+	let coin_gecko_oracle = create_coin_gecko_oracle(Arc::new(OcallApi));
+
+	match get_exchange_rate(trading_pair.clone(), coin_gecko_oracle) {
+		Ok(oc) => extrinsic_calls.push(oc),
```
Comment (murerfel) [inline enclave-runtime/src/lib.rs]: Just a minor issue: I would like to have a descriptive variable name: `opaque_call` here, no abbreviation.

**Review by murerfel — COMMENTED** [review]
Looking good, I think the environment variable approach is pretty simple and clean! Just a few minor improvements I'd like to see.

File: app-libs/exchange-oracle/src/test.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	assert!(bit_usd > 0f32);
+	let dot_to_chf =
+		TradingPair { crypto_currency: "DOT".to_string(), fiat_currency: "CHF".to_string() };
+	let dot_chf = oracle.get_exchange_rate(dot_to_chf).unwrap().0;
+	assert!(dot_chf > 0f32);
+	let bit_to_chf =
+		TradingPair { crypto_currency: "BTC".to_string(), fiat_currency: "CHF".to_string() };
+	let bit_chf = oracle.get_exchange_rate(bit_to_chf).unwrap().0;
+
+	//Ensure that get_exchange_rate return a positive rate
```
Comment (haerdib) [inline app-libs/exchange-oracle/src/test.rs]:
```suggestion
	// Ensure that get_exchange_rate returns a positive rate.
```

File: app-libs/exchange-oracle/src/test.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	assert!(dot_usd > ZERO);
+	//Ensure that the exchange rates' values make sense
```
Comment (haerdib) [inline app-libs/exchange-oracle/src/test.rs]: This comment doesn't really help to understand the assertion. How about 
`// Ensure that the exchange rates values for DOTs / Bitcoins are equal, independent of the the used fiat currency. ` ?

File: service/src/main.rs [hunk 6]
```diff
@@ -575,24 +576,35 @@ fn execute_update_market<E: TeeracleApi>(node_api: &Api<sr25519::Pair, WsRpcClie
...
 	};
 
-	println!("[<] Extrinsic got included into a block. Hash: {:?}\n", extrinsic_hash);
+	// Send the extrinsics to the parentchain and wait for InBlock confirmation.
+	for call in extrinsics.into_iter() {
+		let mut hex_encoded_extrinsic = hex::encode(call.encode());
+		hex_encoded_extrinsic.insert_str(0, "0x");
+
+		println!("[>] Update the exchange rate (send the extrinsic)");
+		let extrinsic_hash = match node_api.send_extrinsic(hex_encoded_extrinsic, XtStatus::InBlock)
```
Comment (haerdib) [inline service/src/main.rs]: Is it not a problem to wait until InBlock for every extrinsic that needs to be sent? (InBlock takes  up to 12 secs for every extrinsic).

File: app-libs/exchange-oracle/Cargo.toml [hunk 7]
```diff
@@ -52,3 +52,4 @@ itp-types = { path = "../../core-primitives/types", default-features = false }
 
 [dev-dependencies]
 itp-test = { path = "../../core-primitives/test" }
+substrate-fixed = { package = "substrate-fixed", git = "https://github.com/encointer/substrate-fixed", tag = "v0.5.9"}
```
Comment (haerdib) [inline app-libs/exchange-oracle/Cargo.toml]:
```suggestion
substrate-fixed = { package = "substrate-fixed", git = "https://github.com/encointer/substrate-fixed", tag = "v0.5.9" }
```

File: app-libs/exchange-oracle/src/coin_market_cap.rs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+	string::{String, ToString},
+	time::Duration,
+};
+use url::Url;
+
+const COINMARKETCAP_URL: &str = "https://pro-api.coinmarketcap.com";
+const COINMARKETCAP_KEY_PARAM: &str = "CMC_PRO_API_KEY";
+const FIAT_CURRENCY_PARAM: &str = "convert_id";
+const CRYPTO_CURRENCY_PARAM: &str = "id";
+const COINMARKETCAP_PATH: &str = "v2/cryptocurrency/quotes/latest"; //For free plan
```
Comment (haerdib) [inline app-libs/exchange-oracle/src/coin_market_cap.rs]: `//For free plan`

What does this comment mean?

File: service/src/main.rs [hunk 9]
```diff
@@ -575,24 +576,35 @@ fn execute_update_market<E: TeeracleApi>(node_api: &Api<sr25519::Pair, WsRpcClie
...
+		{
+			Err(e) => {
+				error!("{:?}: ", e);
+				set_extrinsics_inclusion_success(false);
+				return
+			},
+			Ok(r) => {
```
Comment (haerdib) [inline service/src/main.rs]: `r` -> `hash`

**Review by haerdib — COMMENTED** [review]
One perforamnce question, otherwise only minor comments. Looks very good !

File: service/src/main.rs (hunk 6 above)
Comment (murerfel) [reply service/src/main.rs]: That's a good point - is there an alternative? Can we run this in a non-blocking fashion?

Comment (haerdib) [reply service/src/main.rs]: Couldn't we run asynchronously, and update the metrics upon an `InBlock` event? Or we could also run it within multiple threads, as we are on the untrusted side. What do you think?

Comment (murerfel) [reply service/src/main.rs]: Ah! We're in the untrusted side here, I didn't realize that at first. Then yes, we can use asynchronous patterns. We can use `tokio` to spawn asynchronous tasks here. Pass a `tokio` handle into this method, then in this iteration here, call: 
```
tokio_handle.spawn(move { 
		let mut hex_encoded_extrinsic = hex::encode(call.encode());
		hex_encoded_extrinsic.insert_str(0, "0x");

		println!("[>] Update the exchange rate (send the extrinsic)");
		let extrinsic_hash = match node_api.send_extrinsic(hex_encoded_extrinsic, XtStatus::InBlock) {
                          ...
                };

                println!("[<] Extrinsic got included into a block. Hash: {:?}\n", extrinsic_hash);
 } );
```

Just let me know if you need help with that (the code example might not be exactly correct, I was just freestyling).

File: app-libs/exchange-oracle/src/test.rs (hunk 5 above)
Comment (echevrier) [reply app-libs/exchange-oracle/src/test.rs]: I don't think it helps ;) What about 
Ensure that get_exchange_rate returns a valid value by comparing the value got for DOT/USD and BIT/USD with those for DOT/CHF and BIT/CHF ?

File: app-libs/exchange-oracle/src/coin_market_cap.rs (hunk 8 above)
Comment (echevrier) [reply app-libs/exchange-oracle/src/coin_market_cap.rs]: What about : //API endpoint to get the exchange rate with a basic API plan (free) ?

Comment (haerdib) [reply app-libs/exchange-oracle/src/coin_market_cap.rs]: Is `COINMARKETCAP_PATH` itself not descriptive enough?

File: app-libs/exchange-oracle/src/test.rs (hunk 5 above)
Comment (haerdib) [reply app-libs/exchange-oracle/src/test.rs]: That's exactly what the code states, and doesn't explain what you were thinking when you coded it. What you should add here is why comparing those values makes it valid.

File: app-libs/exchange-oracle/src/coin_market_cap.rs (hunk 8 above)
Comment (echevrier) [reply app-libs/exchange-oracle/src/coin_market_cap.rs]: The focus is on defining the path. There are other API endpoints to get the exchange rate, which however require a paid plan

Comment (haerdib) [reply app-libs/exchange-oracle/src/coin_market_cap.rs]: Ok, then I agree on `// API endpoint to get the exchange rate with a basic API plan (free)`

File: app-libs/exchange-oracle/src/test.rs (hunk 5 above)
Comment (echevrier) [reply app-libs/exchange-oracle/src/test.rs]: What about : Ensure that get_exchange_rate returns a valid value by checking that the values obtained for DOT/BIT from different exchange rates are the same. ?

Comment (haerdib) [reply app-libs/exchange-oracle/src/test.rs]: Yes, I like this one! 👍

**Review by haerdib — APPROVED** [review]
Looks good now. Small remark for the future: Think about the final "." in the comments, but you don't need to let CI rerun for this small thing

**Pull request merged**